File: libs/wasm-sdk/wasm-sdk-zig/src/host.zig
// Host function imports for the Flow-Like WASM runtime.
//
// Each struct groups imports by their WASM module name to avoid Zig-level
// name collisions (e.g. flowlike_vars.get vs flowlike_cache.get).
//
// Public wrapper functions below accept Zig slices and handle pointer
// conversion so callers never deal with raw u32 addresses.

const std = @import("std");
const mem = @import("memory.zig");

// ---------------------------------------------------------------------------
// Raw host imports — flowlike_log
// ---------------------------------------------------------------------------

const log_fns = struct {
    extern "flowlike_log" fn trace(ptr: u32, len: u32) void;
    extern "flowlike_log" fn debug(ptr: u32, len: u32) void;
    extern "flowlike_log" fn info(ptr: u32, len: u32) void;
    extern "flowlike_log" fn warn(ptr: u32, len: u32) void;
    extern "flowlike_log" fn @"error"(ptr: u32, len: u32) void;
    extern "flowlike_log" fn log_json(level: i32, msg_ptr: u32, msg_len: u32, data_ptr: u32, data_len: u32) void;
};

// ---------------------------------------------------------------------------
// Raw host imports — flowlike_pins
// ---------------------------------------------------------------------------

const pin_fns = struct {
    extern "flowlike_pins" fn get_input(name_ptr: u32, name_len: u32) i64;
    extern "flowlike_pins" fn set_output(name_ptr: u32, name_len: u32, val_ptr: u32, val_len: u32) void;
    extern "flowlike_pins" fn activate_exec(name_ptr: u32, name_len: u32) void;
};

// ---------------------------------------------------------------------------
// Raw host imports — flowlike_vars
// ---------------------------------------------------------------------------

const var_fns = struct {
    extern "flowlike_vars" fn get(name_ptr: u32, name_len: u32) i64;
    extern "flowlike_vars" fn set(name_ptr: u32, name_len: u32, val_ptr: u32, val_len: u32) void;
    extern "flowlike_vars" fn delete(name_ptr: u32, name_len: u32) void;
    extern "flowlike_vars" fn has(name_ptr: u32, name_len: u32) i32;
};

// ---------------------------------------------------------------------------
// Raw host imports — flowlike_cache
// ---------------------------------------------------------------------------

const cache_fns = struct {
    extern "flowlike_cache" fn get(key_ptr: u32, key_len: u32) i64;
    extern "flowlike_cache" fn set(key_ptr: u32, key_len: u32, val_ptr: u32, val_len: u32) void;
    extern "flowlike_cache" fn delete(key_ptr: u32, key_len: u32) void;
    extern "flowlike_cache" fn has(key_ptr: u32, key_len: u32) i32;
};

// ---------------------------------------------------------------------------
// Raw host imports — flowlike_meta
// ---------------------------------------------------------------------------

const meta_fns = struct {
    extern "flowlike_meta" fn get_node_id() i64;
    extern "flowlike_meta" fn get_run_id() i64;
    extern "flowlike_meta" fn get_app_id() i64;
    extern "flowlike_meta" fn get_board_id() i64;
    extern "flowlike_meta" fn get_user_id() i64;
    extern "flowlike_meta" fn is_streaming() i32;
    extern "flowlike_meta" fn get_log_level() i32;
    extern "flowlike_meta" fn time_now() i64;
    extern "flowlike_meta" fn random() i64;
};

// ---------------------------------------------------------------------------
// Raw host imports — flowlike_storage
// ---------------------------------------------------------------------------

const storage_fns = struct {
    extern "flowlike_storage" fn read_request(path_ptr: u32, path_len: u32) i64;
    extern "flowlike_storage" fn write_request(path_ptr: u32, path_len: u32, data_ptr: u32, data_len: u32) i32;
    extern "flowlike_storage" fn storage_dir(node_scoped: i32) i64;
    extern "flowlike_storage" fn upload_dir() i64;
    extern "flowlike_storage" fn cache_dir(node_scoped: i32, user_scoped: i32) i64;
    extern "flowlike_storage" fn user_dir(node_scoped: i32) i64;
    extern "flowlike_storage" fn list_request(path_ptr: u32, path_len: u32) i64;
};

// ---------------------------------------------------------------------------
// Raw host imports — flowlike_models
// ---------------------------------------------------------------------------

const model_fns = struct {
    extern "flowlike_models" fn embed_text(bit_ptr: u32, bit_len: u32, texts_ptr: u32, texts_len: u32) i64;
};

// ---------------------------------------------------------------------------
// Raw host imports — flowlike_http
// ---------------------------------------------------------------------------

const http_fns = struct {
    extern "flowlike_http" fn request(method: i32, url_ptr: u32, url_len: u32, headers_ptr: u32, headers_len: u32, body_ptr: u32, body_len: u32) i32;
};

// ---------------------------------------------------------------------------
// Raw host imports — flowlike_stream
// ---------------------------------------------------------------------------

const stream_fns = struct {
    extern "flowlike_stream" fn emit(event_ptr: u32, event_len: u32, data_ptr: u32, data_len: u32) void;
    extern "flowlike_stream" fn text(text_ptr: u32, text_len: u32) void;
};

// ---------------------------------------------------------------------------
// Raw host imports — flowlike_auth
// ---------------------------------------------------------------------------

const auth_fns = struct {
    extern "flowlike_auth" fn get_oauth_token(provider_ptr: u32, provider_len: u32) i64;
    extern "flowlike_auth" fn has_oauth_token(provider_ptr: u32, provider_len: u32) i32;
};

// ---------------------------------------------------------------------------
// Pointer helpers
// ---------------------------------------------------------------------------

inline fn toRaw(s: []const u8) struct { ptr: u32, len: u32 } {
    if (s.len == 0) return .{ .ptr = 0, .len = 0 };
    return .{ .ptr = @intFromPtr(s.ptr), .len = @intCast(s.len) };
}

inline fn unpackString(val: i64) []const u8 {
    return mem.ptrToSlice(mem.unpackHigh(val), mem.unpackLow(val));
}

// ---------------------------------------------------------------------------
// Logging wrappers
// ---------------------------------------------------------------------------

pub fn logTrace(msg: []const u8) void {
    const r = toRaw(msg);
    log_fns.trace(r.ptr, r.len);
}

pub fn logDebug(msg: []const u8) void {
    const r = toRaw(msg);
    log_fns.debug(r.ptr, r.len);
}

pub fn logInfo(msg: []const u8) void {
    const r = toRaw(msg);
    log_fns.info(r.ptr, r.len);
}

pub fn logWarn(msg: []const u8) void {
    const r = toRaw(msg);
    log_fns.warn(r.ptr, r.len);
}

pub fn logError(msg: []const u8) void {
    const r = toRaw(msg);
    log_fns.@"error"(r.ptr, r.len);
}

pub fn logJson(level: i32, msg: []const u8, data: []const u8) void {
    const m = toRaw(msg);
    const d = toRaw(data);
    log_fns.log_json(level, m.ptr, m.len, d.ptr, d.len);
}

// ---------------------------------------------------------------------------
// Pin wrappers
// ---------------------------------------------------------------------------

pub fn getInput(name: []const u8) []const u8 {
    const r = toRaw(name);
    return unpackString(pin_fns.get_input(r.ptr, r.len));
}

pub fn setOutput(name: []const u8, value: []const u8) void {
    const n = toRaw(name);
    const v = toRaw(value);
    pin_fns.set_output(n.ptr, n.len, v.ptr, v.len);
}

pub fn activateExec(name: []const u8) void {
    const r = toRaw(name);
    pin_fns.activate_exec(r.ptr, r.len);
}

// ---------------------------------------------------------------------------
// Variable wrappers
// ---------------------------------------------------------------------------

pub fn getVariable(name: []const u8) []const u8 {
    const r = toRaw(name);
    return unpackString(var_fns.get(r.ptr, r.len));
}

pub fn setVariable(name: []const u8, value: []const u8) void {
    const n = toRaw(name);
    const v = toRaw(value);
    var_fns.set(n.ptr, n.len, v.ptr, v.len);
}

pub fn deleteVariable(name: []const u8) void {
    const r = toRaw(name);
    var_fns.delete(r.ptr, r.len);
}

pub fn hasVariable(name: []const u8) bool {
    const r = toRaw(name);
    return var_fns.has(r.ptr, r.len) != 0;
}

// ---------------------------------------------------------------------------
// Cache wrappers
// ---------------------------------------------------------------------------

pub fn cacheGet(key: []const u8) []const u8 {
    const r = toRaw(key);
    return unpackString(cache_fns.get(r.ptr, r.len));
}

pub fn cacheSet(key: []const u8, value: []const u8) void {
    const k = toRaw(key);
    const v = toRaw(value);
    cache_fns.set(k.ptr, k.len, v.ptr, v.len);
}

pub fn cacheDelete(key: []const u8) void {
    const r = toRaw(key);
    cache_fns.delete(r.ptr, r.len);
}

pub fn cacheHas(key: []const u8) bool {
    const r = toRaw(key);
    return cache_fns.has(r.ptr, r.len) != 0;
}

// ---------------------------------------------------------------------------
// Metadata wrappers
// ---------------------------------------------------------------------------

pub fn getNodeId() []const u8 {
    return unpackString(meta_fns.get_node_id());
}
pub fn getRunId() []const u8 {
    return unpackString(meta_fns.get_run_id());
}
pub fn getAppId() []const u8 {
    return unpackString(meta_fns.get_app_id());
}
pub fn getBoardId() []const u8 {
    return unpackString(meta_fns.get_board_id());
}
pub fn getUserId() []const u8 {
    return unpackString(meta_fns.get_user_id());
}
pub fn isStreaming() bool {
    return meta_fns.is_streaming() != 0;
}
pub fn getLogLevel() i32 {
    return meta_fns.get_log_level();
}
pub fn timeNow() i64 {
    return meta_fns.time_now();
}
pub fn random() i64 {
    return meta_fns.random();
}

// ---------------------------------------------------------------------------
// Storage wrappers
// ---------------------------------------------------------------------------

pub fn storageRead(path: []const u8) []const u8 {
    const r = toRaw(path);
    return unpackString(storage_fns.read_request(r.ptr, r.len));
}

pub fn storageWrite(path: []const u8, data: []const u8) bool {
    const p = toRaw(path);
    const d = toRaw(data);
    return storage_fns.write_request(p.ptr, p.len, d.ptr, d.len) != 0;
}

pub fn storageDir(node_scoped: bool) []const u8 {
    return unpackString(storage_fns.storage_dir(if (node_scoped) @as(i32, 1) else @as(i32, 0)));
}

pub fn uploadDir() []const u8 {
    return unpackString(storage_fns.upload_dir());
}

pub fn cacheDirPath(node_scoped: bool, user_scoped: bool) []const u8 {
    return unpackString(storage_fns.cache_dir(
        if (node_scoped) @as(i32, 1) else @as(i32, 0),
        if (user_scoped) @as(i32, 1) else @as(i32, 0),
    ));
}

pub fn userDir(node_scoped: bool) []const u8 {
    return unpackString(storage_fns.user_dir(if (node_scoped) @as(i32, 1) else @as(i32, 0)));
}

pub fn storageList(flow_path_json: []const u8) []const u8 {
    const r = toRaw(flow_path_json);
    return unpackString(storage_fns.list_request(r.ptr, r.len));
}

// ---------------------------------------------------------------------------
// Model wrappers
// ---------------------------------------------------------------------------

pub fn embedText(bit_json: []const u8, texts_json: []const u8) []const u8 {
    const b = toRaw(bit_json);
    const t = toRaw(texts_json);
    return unpackString(model_fns.embed_text(b.ptr, b.len, t.ptr, t.len));
}

// ---------------------------------------------------------------------------
// HTTP wrappers
// ---------------------------------------------------------------------------

pub fn httpRequest(method: i32, url: []const u8, headers: []const u8, body: []const u8) bool {
    const u = toRaw(url);
    const h = toRaw(headers);
    const b = toRaw(body);
    return http_fns.request(method, u.ptr, u.len, h.ptr, h.len, b.ptr, b.len) != 0;
}

// ---------------------------------------------------------------------------
// Stream wrappers
// ---------------------------------------------------------------------------

pub fn streamEmit(event_type: []const u8, data: []const u8) void {
    const e = toRaw(event_type);
    const d = toRaw(data);
    stream_fns.emit(e.ptr, e.len, d.ptr, d.len);
}

pub fn streamText(txt: []const u8) void {
    const r = toRaw(txt);
    stream_fns.text(r.ptr, r.len);
}

// ---------------------------------------------------------------------------
// Auth wrappers
// ---------------------------------------------------------------------------

pub fn getOAuthToken(provider: []const u8) []const u8 {
    const r = toRaw(provider);
    return unpackString(auth_fns.get_oauth_token(r.ptr, r.len));
}

pub fn hasOAuthToken(provider: []const u8) bool {
    const r = toRaw(provider);
    return auth_fns.has_oauth_token(r.ptr, r.len) != 0;
}

File: libs/wasm-sdk/wasm-sdk-assemblyscript/assembly/host.ts
import { ABI_VERSION, NodeDefinition, ExecutionResult, ExecutionInput, LogLevel } from "./types";
import { parseExecutionInputJson } from "./json";

// -- flowlike_log --

@external("flowlike_log", "trace")
declare function _log_trace(ptr: i32, len: i32): void;
@external("flowlike_log", "debug")
declare function _log_debug(ptr: i32, len: i32): void;
@external("flowlike_log", "info")
declare function _log_info(ptr: i32, len: i32): void;
@external("flowlike_log", "warn")
declare function _log_warn(ptr: i32, len: i32): void;
@external("flowlike_log", "error")
declare function _log_error(ptr: i32, len: i32): void;
@external("flowlike_log", "log_json")
declare function _log_json(level: i32, msg_ptr: i32, msg_len: i32, data_ptr: i32, data_len: i32): void;

// -- flowlike_pins --

@external("flowlike_pins", "get_input")
declare function _get_input(name_ptr: i32, name_len: i32): i64;
@external("flowlike_pins", "set_output")
declare function _set_output(name_ptr: i32, name_len: i32, value_ptr: i32, value_len: i32): void;
@external("flowlike_pins", "activate_exec")
declare function _activate_exec(name_ptr: i32, name_len: i32): void;

// -- flowlike_vars --

@external("flowlike_vars", "get")
declare function _var_get(name_ptr: i32, name_len: i32): i64;
@external("flowlike_vars", "set")
declare function _var_set(name_ptr: i32, name_len: i32, value_ptr: i32, value_len: i32): void;
@external("flowlike_vars", "delete")
declare function _var_delete(name_ptr: i32, name_len: i32): void;
@external("flowlike_vars", "has")
declare function _var_has(name_ptr: i32, name_len: i32): i32;

// -- flowlike_cache --

@external("flowlike_cache", "get")
declare function _cache_get(key_ptr: i32, key_len: i32): i64;
@external("flowlike_cache", "set")
declare function _cache_set(key_ptr: i32, key_len: i32, val_ptr: i32, val_len: i32): void;
@external("flowlike_cache", "delete")
declare function _cache_delete(key_ptr: i32, key_len: i32): void;
@external("flowlike_cache", "has")
declare function _cache_has(key_ptr: i32, key_len: i32): i32;

// -- flowlike_meta --

@external("flowlike_meta", "get_node_id")
declare function _get_node_id(): i64;
@external("flowlike_meta", "get_run_id")
declare function _get_run_id(): i64;
@external("flowlike_meta", "get_app_id")
declare function _get_app_id(): i64;
@external("flowlike_meta", "get_board_id")
declare function _get_board_id(): i64;
@external("flowlike_meta", "get_user_id")
declare function _get_user_id(): i64;
@external("flowlike_meta", "is_streaming")
declare function _is_streaming(): i32;
@external("flowlike_meta", "get_log_level")
declare function _get_log_level(): i32;
@external("flowlike_meta", "time_now")
declare function _time_now(): i64;
@external("flowlike_meta", "random")
declare function _random(): i64;

// -- flowlike_storage --

@external("flowlike_storage", "read_request")
declare function _storage_read(path_ptr: i32, path_len: i32): i64;
@external("flowlike_storage", "write_request")
declare function _storage_write(path_ptr: i32, path_len: i32, data_ptr: i32, data_len: i32): i32;
@external("flowlike_storage", "storage_dir")
declare function _storage_dir(node_scoped: i32): i64;
@external("flowlike_storage", "upload_dir")
declare function _upload_dir(): i64;
@external("flowlike_storage", "cache_dir")
declare function _cache_dir(node_scoped: i32, user_scoped: i32): i64;
@external("flowlike_storage", "user_dir")
declare function _user_dir(node_scoped: i32): i64;
@external("flowlike_storage", "list_request")
declare function _storage_list(path_ptr: i32, path_len: i32): i64;

// -- flowlike_models --

@external("flowlike_models", "embed_text")
declare function _embed_text(bit_ptr: i32, bit_len: i32, texts_ptr: i32, texts_len: i32): i64;

// -- flowlike_http --

@external("flowlike_http", "request")
declare function _http_request(method: i32, url_ptr: i32, url_len: i32, headers_ptr: i32, headers_len: i32, body_ptr: i32, body_len: i32): i32;

// -- flowlike_stream --

@external("flowlike_stream", "emit")
declare function _stream_emit(event_ptr: i32, event_len: i32, data_ptr: i32, data_len: i32): void;
@external("flowlike_stream", "text")
declare function _stream_text(text_ptr: i32, text_len: i32): void;

// -- flowlike_auth --

@external("flowlike_auth", "get_oauth_token")
declare function _get_oauth_token(provider_ptr: i32, provider_len: i32): i64;
@external("flowlike_auth", "has_oauth_token")
declare function _has_oauth_token(provider_ptr: i32, provider_len: i32): i32;

// ============================================================================
// Unpacking helpers for host-returned packed u64 (ptr|len)
// ============================================================================

function unpackString(packed: i64): string | null {
  if (packed == 0) return null;
  const ptr = i32(packed >> 32);
  const len = i32(packed & 0xFFFFFFFF);
  if (ptr == 0 || len == 0) return null;
  const buf = new Uint8Array(len);
  memory.copy(changetype<i32>(buf.buffer), ptr, len);
  return String.UTF8.decode(buf.buffer);
}

function unpackBytes(packed: i64): ArrayBuffer | null {
  if (packed == 0) return null;
  const ptr = i32(packed >> 32);
  const len = i32(packed & 0xFFFFFFFF);
  if (ptr == 0 || len == 0) return null;
  const buf = new ArrayBuffer(len);
  memory.copy(changetype<i32>(buf), ptr, len);
  return buf;
}

// ============================================================================
// Logging
// ============================================================================

export function trace(message: string): void {
  const buf = String.UTF8.encode(message);
  _log_trace(changetype<i32>(buf), buf.byteLength);
}

export function debug(message: string): void {
  const buf = String.UTF8.encode(message);
  _log_debug(changetype<i32>(buf), buf.byteLength);
}

export function info(message: string): void {
  const buf = String.UTF8.encode(message);
  _log_info(changetype<i32>(buf), buf.byteLength);
}

export function warn(message: string): void {
  const buf = String.UTF8.encode(message);
  _log_warn(changetype<i32>(buf), buf.byteLength);
}

export function error(message: string): void {
  const buf = String.UTF8.encode(message);
  _log_error(changetype<i32>(buf), buf.byteLength);
}

export function logJson(level: LogLevel, message: string, data: string): void {
  const msgBuf = String.UTF8.encode(message);
  const dataBuf = String.UTF8.encode(data);
  _log_json(level, changetype<i32>(msgBuf), msgBuf.byteLength, changetype<i32>(dataBuf), dataBuf.byteLength);
}

// ============================================================================
// Pins
// ============================================================================

export function getInput(name: string): string | null {
  const buf = String.UTF8.encode(name);
  const result = _get_input(changetype<i32>(buf), buf.byteLength);
  return unpackString(result);
}

export function setOutput(name: string, value: string): void {
  const nameBuf = String.UTF8.encode(name);
  const valueBuf = String.UTF8.encode(value);
  _set_output(changetype<i32>(nameBuf), nameBuf.byteLength, changetype<i32>(valueBuf), valueBuf.byteLength);
}

export function activateExec(name: string): void {
  const buf = String.UTF8.encode(name);
  _activate_exec(changetype<i32>(buf), buf.byteLength);
}

// ============================================================================
// Variables
// ============================================================================

export function getVariable(name: string): string | null {
  const buf = String.UTF8.encode(name);
  const result = _var_get(changetype<i32>(buf), buf.byteLength);
  return unpackString(result);
}

export function setVariable(name: string, value: string): void {
  const nameBuf = String.UTF8.encode(name);
  const valueBuf = String.UTF8.encode(value);
  _var_set(changetype<i32>(nameBuf), nameBuf.byteLength, changetype<i32>(valueBuf), valueBuf.byteLength);
}

export function deleteVariable(name: string): void {
  const buf = String.UTF8.encode(name);
  _var_delete(changetype<i32>(buf), buf.byteLength);
}

export function hasVariable(name: string): bool {
  const buf = String.UTF8.encode(name);
  return _var_has(changetype<i32>(buf), buf.byteLength) != 0;
}

// ============================================================================
// Cache
// ============================================================================

export function cacheGet(key: string): string | null {
  const buf = String.UTF8.encode(key);
  const result = _cache_get(changetype<i32>(buf), buf.byteLength);
  return unpackString(result);
}

export function cacheSet(key: string, value: string): void {
  const keyBuf = String.UTF8.encode(key);
  const valBuf = String.UTF8.encode(value);
  _cache_set(changetype<i32>(keyBuf), keyBuf.byteLength, changetype<i32>(valBuf), valBuf.byteLength);
}

export function cacheDelete(key: string): void {
  const buf = String.UTF8.encode(key);
  _cache_delete(changetype<i32>(buf), buf.byteLength);
}

export function cacheHas(key: string): bool {
  const buf = String.UTF8.encode(key);
  return _cache_has(changetype<i32>(buf), buf.byteLength) != 0;
}

// ============================================================================
// Metadata
// ============================================================================

export function getNodeId(): string {
  const result = unpackString(_get_node_id());
  return result !== null ? result : "";
}

export function getRunId(): string {
  const result = unpackString(_get_run_id());
  return result !== null ? result : "";
}

export function getAppId(): string {
  const result = unpackString(_get_app_id());
  return result !== null ? result : "";
}

export function getBoardId(): string {
  const result = unpackString(_get_board_id());
  return result !== null ? result : "";
}

export function getUserId(): string {
  const result = unpackString(_get_user_id());
  return result !== null ? result : "";
}

export function isStreaming(): bool {
  return _is_streaming() != 0;
}

export function getLogLevel(): i32 {
  return _get_log_level();
}

export function now(): i64 {
  return _time_now();
}

export function random(): i64 {
  return _random();
}

// ============================================================================
// Storage
// ============================================================================

export function storageRead(path: string): ArrayBuffer | null {
  const buf = String.UTF8.encode(path);
  const result = _storage_read(changetype<i32>(buf), buf.byteLength);
  return unpackBytes(result);
}

export function storageWrite(path: string, data: ArrayBuffer): bool {
  const pathBuf = String.UTF8.encode(path);
  return _storage_write(changetype<i32>(pathBuf), pathBuf.byteLength, changetype<i32>(data), data.byteLength) != 0;
}

export function storageDir(nodeScoped: bool): string | null {
  return unpackString(_storage_dir(nodeScoped ? 1 : 0));
}

export function uploadDir(): string | null {
  return unpackString(_upload_dir());
}

export function cacheDir(nodeScoped: bool, userScoped: bool): string | null {
  return unpackString(_cache_dir(nodeScoped ? 1 : 0, userScoped ? 1 : 0));
}

export function userDir(nodeScoped: bool): string | null {
  return unpackString(_user_dir(nodeScoped ? 1 : 0));
}

export function storageList(flowPathJson: string): string | null {
  const buf = String.UTF8.encode(flowPathJson);
  const result = _storage_list(changetype<i32>(buf), buf.byteLength);
  return unpackString(result);
}

// ============================================================================
// Models
// ============================================================================

export function embedText(bitJson: string, textsJson: string): string | null {
  const bitBuf = String.UTF8.encode(bitJson);
  const textsBuf = String.UTF8.encode(textsJson);
  const result = _embed_text(
    changetype<i32>(bitBuf), bitBuf.byteLength,
    changetype<i32>(textsBuf), textsBuf.byteLength
  );
  return unpackString(result);
}

// ============================================================================
// HTTP
// ============================================================================

export function httpRequest(method: i32, url: string, headers: string, body: string): bool {
  const urlBuf = String.UTF8.encode(url);
  const headersBuf = String.UTF8.encode(headers);
  const bodyBuf = String.UTF8.encode(body);
  return _http_request(
    method,
    changetype<i32>(urlBuf), urlBuf.byteLength,
    changetype<i32>(headersBuf), headersBuf.byteLength,
    changetype<i32>(bodyBuf), bodyBuf.byteLength
  ) != 0;
}

// ============================================================================
// Streaming
// ============================================================================

export function stream(eventType: string, data: string): void {
  const typeBuf = String.UTF8.encode(eventType);
  const dataBuf = String.UTF8.encode(data);
  _stream_emit(changetype<i32>(typeBuf), typeBuf.byteLength, changetype<i32>(dataBuf), dataBuf.byteLength);
}

export function streamText(text: string): void {
  const buf = String.UTF8.encode(text);
  _stream_text(changetype<i32>(buf), buf.byteLength);
}

export function streamProgress(progress: f32, message: string): void {
  stream("progress", `{"progress":${progress},"message":"${message}"}`);
}

// ============================================================================
// Auth
// ============================================================================

export function getOAuthToken(provider: string): string | null {
  const buf = String.UTF8.encode(provider);
  const result = _get_oauth_token(changetype<i32>(buf), buf.byteLength);
  return unpackString(result);
}

export function hasOAuthToken(provider: string): bool {
  const buf = String.UTF8.encode(provider);
  return _has_oauth_token(changetype<i32>(buf), buf.byteLength) != 0;
}

// ============================================================================
// Memory Exports
// ============================================================================

let resultBuffer: ArrayBuffer = new ArrayBuffer(0);

export function alloc(size: i32): i32 {
  const buf = new ArrayBuffer(size);
  return changetype<i32>(buf);
}

export function dealloc(_ptr: i32, _size: i32): void {}

export function get_abi_version(): i32 {
  return ABI_VERSION;
}

export function packResult(json: string): i64 {
  const buf = String.UTF8.encode(json);
  resultBuffer = buf;
  const ptr = changetype<i32>(buf);
  const len = buf.byteLength;
  return (i64(ptr) << 32) | i64(len);
}

export function serializeDefinition(def: NodeDefinition): i64 {
  return packResult(def.toJSON());
}

export function serializeResult(result: ExecutionResult): i64 {
  return packResult(result.toJSON());
}

export function parseInput(ptr: i32, len: i32): ExecutionInput {
  const buf = new Uint8Array(len);
  memory.copy(changetype<i32>(buf.buffer), ptr, len);
  const json = String.UTF8.decode(buf.buffer);
  return parseExecutionInputJson(json);
}
